feat: add rate limiting provider option

PendaGTP opened the pull request:
# Description

Add OVH API rate limiting option

Currently, the rate-limiting is controlled by `GOMAXPROCS,` and if you have many threads, you will exceed the rate limit of the OVH API (e.g https://github.com/kubernetes-sigs/external-dns/issues/1546). This adds a simple rate-limiting implementation similar to what was implemented in external-dns (https://github.com/kubernetes-sigs/external-dns/pull/1619).

Fixes: #980

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation update

# How Has This Been Tested?

Tests are green.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings or issues
- [ ] I have added acceptance tests that prove my fix is effective or that my feature works
- [x] New and existing acceptance tests pass locally with my changes
- [x] I ran succesfully `go mod vendor` if I added or modify `go.mod` file

amstuta reviewed (commented):
Thanks for your contribution, this will be a great feature to have in the provider :)

PendaGTP commented:
@amstuta  Thanks for your feedback! I've made the following changes:

1. Extracted the OVH wrapper into a new package as suggested
2. Refactored the helper functions to accept the rate-limited client
3. Rebased this branch with upstream to resolve any conflicts

Let me know if these changes address your concerns!

amstuta commented:
Thanks @PendaGTP, LGTM :) 

I'm just wondering one last thing, could we remove the `RawOVHClient` from the config ?

PendaGTP commented:
@amstuta Done! Removed `RawOVHClient` from config and embedded `*ovh.Client` directly in the wrapper. Let me know if there aree any other changes required or any other suggestions!

amstuta commented:
I just tested your branch locally and it works quite well !

I just see two remaining things to fix before we can merge:

- You should add a block like the following in [here](https://github.com/ovh/terraform-provider-ovh/blob/master/ovh/provider.go#L299), otherwise the provider will panic if no rate-limit is defined:
```go
if v, ok := d.GetOk("api_rate_limit"); ok {
    config.ApiRateLimit = ratelimit.New(v.(int))
} else {
    config.ApiRateLimit = ratelimit.NewUnlimited()
}
```

- Add this change: https://github.com/ovh/terraform-provider-ovh/pull/979/files#diff-b4d68dc855d0f9476d3f2ee343853bd21bf82ea9960d0cf06661baa244439dd6 also in the [template file](https://github.com/ovh/terraform-provider-ovh/blob/master/templates/index.md.tmpl)

After that, ok for me to merge :)

PendaGTP commented:
Thanks for testing and helping me implement this feature properly! I've implemented both requested changes.
Everything should be ready for merge now. Let me know if you'd like me to rebase or squash the commits before merging.
